Add Armeria version metric

matsumana opened the pull request:
Hello, I've added Armeria version metric.

**Motivation:**
- Want to check Armeria version with Grafana
  FYI:
  In Prometheus best practices, this kind of metric is described as follows.
  https://prometheus.io/docs/practices/naming/
  > foobar_build_info (for a pseudo-metric that provides metadata about the running binary)
  
  e.g.
  ```
  # HELP node_exporter_build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which node_exporter was built.
  # TYPE node_exporter_build_info gauge
  node_exporter_build_info{branch="master",goversion="go1.7.5",revision="840ba5dcc71a084a3bc63cb6063003c1f94435a6",version="0.14.0"} 1
  ```

**Modifications:**

- Add Armeria version metric

**Result:**

It's exported as follows

Labels:
- commit
- version

```
# HELP armeria_build_info A metric with a constant '1' value labeled by version and commit hash from which Armeria was built.
# TYPE armeria_build_info gauge
armeria_build_info{commit="74700d402178e00045b1ca23006f7f6131b2a92d",version="0.94.1-SNAPSHOT",} 1.0
```

example dashboard:
![](https://matsumana.files.wordpress.com/2019/10/screen-shot-2019-10-10-at-22.35.12.png)

trustin commented:
That is awesome, @matsumana. I love it! What do you think about adding other labels such as `repoStatus` whose value is `clean` or `dirty`?

anuraaga commented on core/src/test/java/com/linecorp/armeria/server/ServerTest.java:
```diff
@@ -436,6 +436,18 @@ public void notGracefulShutdownBlockingTaskExecutor() {
+        final MeterRegistry meterRegistry = server.config().meterRegistry();
+        assertThat(meterRegistry.find("armeria.build.info")
+                                .gauge()
+                                .value())
```
How about also checking the three tags exist (though not checking their values which isn't constant)

matsumana commented:
> What do you think about adding other labels such as `repoStatus` whose value is `clean` or `dirty`?

Sounds good.
Added!

```
# HELP armeria_build_info A metric with a constant '1' value labeled by version and commit hash from which Armeria was built.
# TYPE armeria_build_info gauge
armeria_build_info{commit="74700d402178e00045b1ca23006f7f6131b2a92d",repostatus="dirty",version="0.94.1-SNAPSHOT",} 1.0
```

I added `repostatus`, not `repoStatus`.
I followed node_exporter's label naming.
e.g.
`node_filesystem_free{device="/dev/sda2",fstype="ext4",mountpoint="/"} 8.4968587264e+10`

trustin commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -305,6 +311,27 @@ public int numConnections() {
+        final String repositoryStatus = versionInfo.repositoryStatus();
+        final List<Tag> tags = ImmutableList.of(Tag.of("version", version),
+                                                Tag.of("commit", commit),
+                                                Tag.of("repostatus", repositoryStatus));
```
`repoStatus`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/Server.java:
```diff
@@ -305,6 +311,27 @@ public int numConnections() {
+        Gauge.builder("armeria.build.info", () -> 1)
+             .tags(tags)
+             .description("A metric with a constant '1' value labeled by version and commit hash"
+                          + " from which Armeria was built.")
```
Armeria uses a slightly different convention for infix operators with line wrapping.
```java
             .description("A metric with a constant '1' value labeled by version and commit hash" +
                          " from which Armeria was built.")
```

trustin commented:
> I added repostatus, not repoStatus.
> I followed node_exporter's label naming.

Oh, just found this comment. Armeria currently is using camel case convention for label names, e.g. `hostnamePattern`, so I changed it to `repoStatus`. Do you think we should make them all lower-cased?

matsumana commented:
> Oh, just found this comment. Armeria currently is using camel case convention for label names, e.g. `hostnamePattern`, so I changed it to `repoStatus`. Do you think we should make them all lower-cased?

It was my misunderstanding.
I remembered that you guys were discussing about metrics naming conventions in a issue/PR before, so I misunderstood that you are going to migrate from camelCase.

Thank you for changing the label.
Could you merge?

trustin commented:
Ah, it was about meter names rather than label names :-) Will merge soon!

trustin commented:
Much appreciated, @matsumana ! :+1:
